### PR TITLE
Expand docs and add some Default impl's for image

### PR DESCRIPTION
**Pull request by icefoxen**
Fixes https://github.com/amethyst/rendy/issues/117
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: vulkan
- [x] `rustfmt` run on changed code

### CODE REVIEW

File: src/hal/src/image.rs [hunk 1]
```diff
@@ -56,15 +56,20 @@ impl Extent {
...
+/// copy operations.  All offsets are in texels, and
+/// specifying offsets other than 0 for dimensions
+/// that do not exist is undefined behavior -- for
+/// example, specifying a `z` offset of `1` in a
+/// two-dimensional image.
 #[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct Offset {
-    ///
+    /// X offset, in texels
```
Comment (kvark) [inline src/hal/src/image.rs]: I think we can have a `type TexelCoordinate = i32` to reflect this

File: src/hal/src/image.rs [hunk 2]
```diff
@@ -208,6 +213,13 @@ pub enum Filter {
     Linear,
 }
 
+impl Default for Filter {
```
Comment (kvark) [inline src/hal/src/image.rs]: why do we want a default filter here?

File: src/hal/src/image.rs [hunk 3]
```diff
@@ -469,7 +490,10 @@ impl Into<[f32; 4]> for PackedColor {
     }
 }
 
-/// Specifies how to sample from an image.
+/// Specifies how to sample from an image.  This is all the parameters
```
Comment (kvark) [inline src/hal/src/image.rs]: This is -> These are

File: src/hal/src/image.rs (hunk 2 above)
Comment (icefoxen) [reply src/hal/src/image.rs]: Just to impl `Default` for `SamplerInfo`, which is something I wanted recently to make basic hello-world style setup just a little simpler.

File: src/hal/src/image.rs (hunk 3 above)
Comment (icefoxen) [reply src/hal/src/image.rs]: :+1:

**Comment by icefoxen** [comment]
I recently wanted the defaults for writing something that was basically a hello-world program, where I was saying "okay I don't care what values these are yet, I just want them to be something vaguely sensible".  Up to you guys whether or not you actually want them.

File: src/hal/src/image.rs (hunk 1 above)
Comment (icefoxen) [reply src/hal/src/image.rs]: :+1:

**Comment by kvark** [comment]
Problem is, in gfx-hal the goal of reducing user's boilerplate is quite low priority. Cleanness and soundness of the semantics matter most. I would suggest leaving the opinionated defaults to other wrapping higher level libraries (ggez?).

**Comment by icefoxen** [comment]
> Problem is, in gfx-hal the goal of reducing user's boilerplate is quite low priority.

That's why I'm doing it for you.  :-P  But, if you want to leave those sorts of things out on purpose, just confirm it and I'll update the PR's to remove the `Default` impl's.

**Comment by kvark** [comment]
Yes, please. I confirm we don't want gfx-hal to be opinionated about what the user wants. Defaults are accepted when they are for disabling an optional function block, i.e. default for a depth test is OFF, default for any option<> is fine to have None, etc. This is not the case for `SamplerState`.

**Comment by icefoxen** [comment]
Makes sense.  Done!

**Comment by kvark** [comment]
Thank you! Please squash a bit (adding + removing something shouldn't be in history within the same PR)

**Comment by icefoxen** [comment]
Done.

**Pull request merged**